File: contests/lc/955.DeleteColumnstoMakeSortedII/solution.py
from typing import List


class Solution:
    def minDeletionSize(self, strs: List[str]) -> int:
        n = len(strs)
        m = len(strs[0])
        fixed = [False] * (n -1)
        deletions = 0

        for col in range(m):
            bad = False
            for i in range(n -1):
                if not fixed[i] and strs[i][col] > strs[i+1][col]:
                    bad = True
                    break
            if bad:
                deletions += 1
                continue

            for i in range(n-1):
                if not fixed[i] and strs[i][col] < strs[i+1][col]:
                    fixed[i] = True
            if all(fixed):
                break
        return deletions

File: contests/lc/955.DeleteColumnstoMakeSortedII/solution.go
package _55_DeleteColumnstoMakeSortedII

func minDeletionSize(strs []string) int {
	n := len(strs)
	m := len(strs[0])

	fixed := make([]bool, n-1)
	deletions := 0

	allFixed := func() bool {
		for _, v := range fixed {
			if !v {
				return false
			}
		}
		return true
	}

	for col := 0; col < m; col++ {
		bad := false
		for i := 0; i < n-1; i++ {
			if !fixed[i] && strs[i][col] > strs[i+1][col] {
				bad = true
				break
			}
		}
		if bad {
			deletions++
			continue
		}

		for i := 0; i < n-1; i++ {
			if !fixed[i] && strs[i][col] < strs[i+1][col] {
				fixed[i] = true
			}
		}
		if allFixed() {
			break
		}

	}
	return deletions
}
